src/vision/status_quo/alan_builds_a_task_scheduler.md

```diff
@@ -0,0 +1,704 @@
+            notifier: Arc::new(notify::Notifier::new()),
+            ...
+        }
+    }
+    ...
+
+    pub async fn start(&self) -> io::Result<()> {
+        task::spawn(apply_async(
+            self.notifier.clone(),
+            ...
+        ));
+        ...
+        Ok(())
+    }
+    ...
+
+
+pub async fn apply_async(
+    notifier: Arc<notify::Notifier>,
+    ...
+) -> io::Result<()> {
+  ...
+```
+
+Alan is satisfied with the compilation success for the moment, but doesn't feel confident about the production readiness of Rust async.
+
+### Implementing HTTP server handler
+
+Familiar with Go standard libraries, Alan implemented the following request handler without any third-party dependencies:
+
+```go
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"os"
+	"os/signal"
+	"syscall"
+	"time"
+)
+
+type Handler interface {
+	start()
+}
+
+type handler struct {
+	listenerPort uint64
+	applier      Applier
+}
+
+func (hd *handler) start() {
+	hd.applier.start()
+
+	serverMux := http.NewServeMux()
+	serverMux.HandleFunc("/echo", hd.wrapFunc(handleRequest))
+
+	httpServer := &http.Server{
+		Addr:    fmt.Sprintf(":%d", hd.listenerPort),
+		Handler: serverMux,
+	}
+
+	tch := make(chan os.Signal, 1)
+	signal.Notify(tch, syscall.SIGINT)
+	done := make(chan struct{})
+	go func() {
+		httpServer.Close()
+		close(done)
+	}()
+
+	if err := httpServer.ListenAndServe(); err != nil {
+		fmt.Printf("http server error: %v\n", err)
+	}
+	select {
+	case <-done:
+	default:
+	}
+
+	if err := hd.applier.stop(); err != nil {
+		panic(err)
+	}
+}
+
+func (hd *handler) wrapFunc(fn func(applier Applier, w http.ResponseWriter, req *http.Request)) func(w http.ResponseWriter, req *http.Request) {
+	return func(w http.ResponseWriter, req *http.Request) {
+		fn(hd.applier, w, req)
+	}
+}
+
+func handleRequest(applier Applier, w http.ResponseWriter, req *http.Request) {
+	switch req.Method {
+	case "POST":
+		var echoRequest EchoRequest
+		err := json.NewDecoder(req.Body).Decode(&echoRequest)
+		if err != nil {
+			fmt.Fprintf(w, "failed to read request %v", err)
+			return
+		}
+		s, err := applier.apply(Request{echoRequest: &echoRequest})
+		if err != nil {
+			fmt.Fprintf(w, "failed to apply request %v", err)
+			return
+		}
+		fmt.Fprint(w, s)
+
+	default:
+		http.Error(w, "Method Not Allowed", 405)
+	}
+}
+```
+
+For Rust, Alan is faced with multiple choices again:
+
+- [hyper](https://github.com/hyperium/hyper)
+- [actix-web](https://github.com/actix/actix-web)
+- [warp](https://github.com/seanmonstar/warp)
+- [rocket](https://github.com/SergioBenitez/Rocket)
+- [tide](https://github.com/http-rs/tide)
+
+Alan strongly believes in Go's minimal dependency approach, and thereby chooses "hyper" for its low-level API. While "hyper" is meant to be a low-level building block, implementing a simple request handler in "hyper" still requires four different external dependencies. Alan is not surprised anymore, and rather accepts the status quo of split Rust ecosystem:
```

Comment:
I am surprised by this since hyper would be my last choice. hyper already mentioned that they are "low-level" and therefore is not recommended for normal use. Did you faced the same situation in the past but still chose hyper? Can I know if having a minimum amount of dependency was the reason? But yes, I agree that there is no single official comparison or a guide to choosing rust frameworks is a bad thing, things would have been easier if there is sort of feature comparison and review between them to compare, cargo-crev might help but it does not provide comparison.